core/src/main/java/com/linecorp/armeria/common/stream/AbstractMulticastStreamFactory.java

```diff
@@ -0,0 +1,596 @@
+                if (state != State.OPEN) {
+                    downstreamSubscriptions.remove(subscription);
+                    throw new IllegalStateException("This factory has been closed.");
+                }
+            }
+
+            final Subscriber<Object> subscriber = subscription.subscriber();
+            if (upstreamSubscription != null && subscription.updateOnSubscribe()) {
+                final Executor executor = subscription.executor();
+                if (executor != null) {
+                    executor.execute(() -> subscriber.onSubscribe(subscription));
+                } else {
+                    subscriber.onSubscribe(subscription);
+                }
+            }
+        }
+
+        void requestDemand(long cumulativeDemand) {
+            for (;;) {
+                if (cumulativeDemand <= requestedDemand) {
+                    return;
+                }
+                long currentRequested = requestedDemand;
+                if (requestedDemandUpdater.compareAndSet(this, currentRequested, cumulativeDemand)) {
+                    upstreamSubscription.request(cumulativeDemand - currentRequested);
+                    return;
+                }
+            }
+        }
+
+        void notifyDownstreams() {
+            if (downstreamSubscriptions.isEmpty()) {
+                return;
+            }
+            downstreamSubscriptions.forEach(downstream -> {
+                final Executor executor = downstream.executor();
+                if (executor != null) {
+                    executor.execute(() -> notifyDownstream(downstream));
+                } else {
+                    notifyDownstream(downstream);
+                }
+            });
+        }
+
+        void notifyDownstream(DownstreamSubscription downstream) {
+            for (;;) {
+                int offsetOfSubscriber = downstream.offset;
+                int upstreamOffset = this.upstreamOffset;
+                if (upstreamCompleted && offsetOfSubscriber == upstreamOffset) {
+                    if (DownstreamSubscription.writeFlagUpdater.compareAndSet(downstream, 0, 1)) {
+                        completeDownstream(downstream, upstreamCause);
+                    }
+                    return;
+                }
+
+                if (downstream.cancelled) {
+                    if (DownstreamSubscription.writeFlagUpdater.compareAndSet(downstream, 0, 1)) {
+                        completeDownstream(downstream, CancelledSubscriptionException.get());
+                    }
+                    downstreamSubscriptions.remove(downstream);
+                    return;
+                }
+
+                if (offsetOfSubscriber == upstreamOffset) {
+                    break;
+                }
+
+                if (!publishData(downstream)) {
+                    break;
+                }
+            }
+        }
+
+        private void completeDownstream(DownstreamSubscription downstream, Throwable cause) {
```

Comment:
How about moving `downstreamSubscriptions.remove()` above to this method so that the downstream subscription is always removed on completion?